File: backend/app/api/pdelta.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List
import numpy as np
from app.engine.pdelta import PDeltaAnalysis

router = APIRouter()

class PDeltaRequest(BaseModel):
    stiffness_matrix: List[List[float]]
    mass_matrix: List[List[float]]
    loads: List[float]
    axial_forces: List[float]
    element_lengths: List[float]

class StabilityIndexRequest(BaseModel):
    story_shear: float
    story_weight: float
    story_drift: float
    story_height: float

@router.post("/analysis")
def pdelta_analysis(request: PDeltaRequest):
    """Perform P-Delta analysis"""
    try:
        analyzer = PDeltaAnalysis()
        
        K = np.array(request.stiffness_matrix)
        M = np.array(request.mass_matrix)
        P = np.array(request.loads)
        axial = np.array(request.axial_forces)
        lengths = np.array(request.element_lengths)
        
        result = analyzer.pdelta_analysis(K, M, P, axial, lengths)
        
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/stability-index")
def calculate_stability_index(request: StabilityIndexRequest):
    """Calculate stability index (theta)"""
    try:
        analyzer = PDeltaAnalysis()
        result = analyzer.stability_index(
            request.story_shear,
            request.story_weight,
            request.story_drift,
            request.story_height
        )
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/moment-amplification")
def moment_amplification(Pu: float, Pc: float, Cm: float = 1.0):
    """Calculate moment amplification factor"""
    try:
        analyzer = PDeltaAnalysis()
        result = analyzer.moment_amplification_factor(Pu, Pc, Cm)
        return {"status": "success", "results": result}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/lib/api/pdelta.ts
import { apiClient } from './client';
import type { ApiResponse } from '@/types/api';

export interface PDeltaRequest {
  model_id: number;
  load_case: string;
  max_iterations?: number;
  tolerance?: number;
}

export interface StabilityIndexRequest {
  story_shear: number;
  story_weight: number;
  story_drift: number;
  story_height: number;
}

export const pdeltaApi = {
  runAnalysis: async (data: PDeltaRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/pdelta/analysis', data);
    return response.data.data!;
  },

  calculateStabilityIndex: async (data: StabilityIndexRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/pdelta/stability-index', data);
    return response.data.data!;
  },

  momentAmplification: async (Pu: number, Pc: number, Cm: number = 1.0): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/pdelta/moment-amplification', { Pu, Pc, Cm });
    return response.data.data!;
  },
};
